File: VB/exportPanel/Controllers/HomeController.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Threading
Imports System.Web
Imports System.Web.Mvc
Imports DevExpress.Web.Mvc

Namespace exportPanel.Controllers
	Public Class HomeController
		Inherits Controller
		Public Function Index() As ActionResult
			Return View()
		End Function

		Public Function GridViewPartial() As ActionResult
			Return PartialView(GetData())
		End Function
		Public Function Export() As ActionResult
			Dim dataStream = TryCast(Session("stream"), MemoryStream)
			Dim disposition As String = "attachment"
			Dim response = HttpContext.Response
			response.Clear()
			response.Buffer = False
			response.AppendHeader("Content-Type", String.Format("application/{0}", "pdf"))
			response.AppendHeader("Content-Transfer-Encoding", "binary")
			response.AppendHeader("Content-Disposition", String.Format("{0}; filename={1}.{2}", disposition, HttpUtility.UrlEncode("grid").Replace("+", "%20"), "pdf"))
			response.BinaryWrite(dataStream.ToArray())
			response.End()
			Return Nothing
		End Function
		Public Function CallbackExport() As ActionResult
			Thread.Sleep(5000)
			Dim stream As New MemoryStream()
			GridViewExtension.WritePdf(GetSettings(), GetData(), stream)
			Session("stream") = stream
			Return Nothing
		End Function

		Public Shared Function GetData() As Object
			Return Enumerable.Range(0, 10).Select(Function(i) New With {Key .ID = i, Key .Text = "Text " & i})
		End Function

		Public Shared Function GetSettings() As GridViewSettings
			Dim settings = New GridViewSettings()
			settings.Name = "gridView"
			settings.CallbackRouteValues = New With {Key .Controller = "Home", Key .Action = "GridViewPartial"}
			Return settings
		End Function
	End Class
End Namespace

File: CS/exportPanel/Controllers/HomeController.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading;
using System.Web;
using System.Web.Mvc;
using DevExpress.Web.Mvc;

namespace exportPanel.Controllers {
    public class HomeController : Controller {
        public ActionResult Index() {
            return View();
        }

        public ActionResult GridViewPartial() {
            return PartialView(GetData());
        }
        public ActionResult Export() {
            var dataStream = Session["stream"] as MemoryStream;
            string disposition = "attachment";
            var response = HttpContext.Response;
            response.Clear();
            response.Buffer = false;
            response.AppendHeader("Content-Type", string.Format("application/{0}", "pdf"));
            response.AppendHeader("Content-Transfer-Encoding", "binary");
            response.AppendHeader("Content-Disposition", string.Format("{0}; filename={1}.{2}", disposition, HttpUtility.UrlEncode("grid").Replace("+", "%20"), "pdf"));
            response.BinaryWrite(dataStream.ToArray());
            response.End();
            return null;
        }
        public ActionResult CallbackExport() {
            Thread.Sleep(5000);
            MemoryStream stream = new MemoryStream();
            GridViewExtension.WritePdf(GetSettings(), GetData(), stream);
            Session["stream"] = stream;
            return null;
        }

        public static object GetData() {
            return Enumerable.Range(0, 10).Select(i => new { ID = i, Text = "Text " + i});
        }

        public static GridViewSettings GetSettings() {
            var settings = new GridViewSettings();
            settings.Name = "gridView";
            settings.CallbackRouteValues = new { Controller = "Home", Action = "GridViewPartial" };
            return settings;
        }
    }
}